src/D2L.CodeStyle.Analyzers/Async/Generator/FileCollector.cs

```diff
@@ -0,0 +1,273 @@
+			= SyntaxFactory.Token( SyntaxKind.None );
+
+		private static readonly SyntaxTriviaList EmptyTrivia
+			= SyntaxFactory.TriviaList();
+
+		private readonly CompilationUnitSyntax m_root;
+
+		// This is a mutable dictionary because we will remove from it as we go
+		// and check that it is empty at the end, just to be safe.
+		private readonly Dictionary<SyntaxNode, IEnumerable<string>> m_methods;
+
+		// When we recurse into a namespace or type, we're not sure if the
+		// preamble ("namespace foo {", "\t\tpartial class Bar<T> {", etc.) is
+		// necessary, because maybe there are no generated methods.
+		// Using a LinkedList<string> as a double-ended queue.
+		private readonly LinkedList<string> m_preambles = new();
+
+		private readonly StringBuilder m_out = new();
+
+		private FileCollector(
+			CompilationUnitSyntax root,
+			Dictionary<SyntaxNode, IEnumerable<string>> methods
+		) {
+			m_root = root;
+			m_methods = methods;
+		}
+
+		public static FileCollector Create(
+			CompilationUnitSyntax root,
+			ImmutableArray<(SyntaxNode Parent, string Source)> methods
+		) {
+			var groupedMethods = methods
+				.GroupBy( static m => m.Parent )
+				.ToDictionary(
+					static g => g.Key,
+					static g => g.Select( static m => m.Source )
+				);
+
+			return new FileCollector( root, groupedMethods );
+		}
+
+		public string CollectSource() {
+			// File-scoped usings:
+			m_out.Append( m_root.Usings.ToFullString() );
+
+			WriteChildren( m_root );
+
+			if( m_preambles.Count != 0 ) {
+				throw new Bug( "left over preambles" );
+			}
+
+			if( m_methods.Count != 0 ) {
+				throw new Bug( "left over methods" );
+			}
+
+			return m_out.ToString();
+		}
+
+		private bool WriteChildren( SyntaxNode node ) {
+			// Early-out if we have no methods left to place.
+			if( m_methods.Count == 0 ) {
+				return false;
+			}
+
+			var preambleCountBefore = m_preambles.Count;
+
+			bool wroteSomething = false;
+
+			// First recurse into child types
+			foreach( var child in node.ChildNodes() ) {
+				bool wroteSomethingHere = child switch {
+					ClassDeclarationSyntax @class => WriteType( @class ),
+					StructDeclarationSyntax @struct => WriteType( @struct ),
+					InterfaceDeclarationSyntax @interface => WriteType( @interface ),
+
+					// Covers record + record struct:
+					RecordDeclarationSyntax @record => WriteType( @record, @record.ClassOrStructKeyword ),
+
+					// Covers file-scoped and block namespaces
+					BaseNamespaceDeclarationSyntax @namespace => WriteNamespace( @namespace ),
+
+					// Ignore any other child node
+					_ => false
+				};
+
+				if( wroteSomethingHere ) {
+					wroteSomething = true;
+				}
+
+				CheckPremableInvariant( wroteSomething, preambleCountBefore );
+			}
+
+			return wroteSomething;
+		}
+
+		private bool WriteNamespace( BaseNamespaceDeclarationSyntax @namespace ) {
+			SyntaxToken? closeBrace;
+
+			switch( @namespace ) {
+				case FileScopedNamespaceDeclarationSyntax ns:
+					m_preambles.AddLast(

```

Comment:
Hmm.  Should m_preambles be a `LinkedList<SyntaxToken[]>` to punt the string concatenation down to StringBuilder?

Reply:
Good call. There are two other types beyond `SyntaxToken` that we put into the preamble (covered by `SyntaxNode` and, annoyingly, `SyntaxList<UsingDeclarationSyntax>`). I added a little wrapper struct for this to defer the calls to `ToFullString()`.

There is the win, as you say, from sending them through `StringBuilder`. We also will win a bit for discarded preambles although I suspect that will not be the common case. We pay a tax of allocating that (small) array in gen0 and a little bit of reflection (casts) to implement `ToString()` in the wrapper. This seems reasonable to me.

Something I didn't mention but once we have some serious usage of this it would be good to spy on what the typical size of the `StringBuilder` in practice so that we can tune the initial capacity.

Reply:
Yeah.  Seeing the pooling stuff like this in the Roslyn repo is eye opening